### ISSUE TITLE
The val error does not decrease.

### ISSUE THREAD

**Issue by wlw208dzy** [yueatsprograms/Stochastic_Depth#2]
Thanks for releasing the code. I am using one TitanX to reproduce the results of Stochastic Depth for cifar10, but the val error does not decrease since epoch 1, as shown below:
  resultFolder : "results-cifar10/"
  device : 2
  dataset : "cifar10"
  N : 18
  batchSize : 128
  augmentation : true
  maxEpochs : 500
  deathRate : 0.5
  deathMode : "lin_decay"
}
Training set size:      45000  
Validation set size:        5000  
Test set size:          10000  
Building model...  
Training...
Epoch   Valid. err  Test err    Training time  
Epoch 1:    89.74%      90.00%      51s 
Epoch 2:    90.02%      90.00%      51s 
Epoch 3:    89.74%      90.00%      51s 
Epoch 4:    89.74%      90.00%      51s 
Epoch 5:    89.74%      90.00%      51s 
Epoch 6:    89.74%      90.00%      51s 
Epoch 7:    89.74%      90.00%      51s 
Epoch 8:    89.74%      90.00%      51s 
Epoch 9:    89.74%      90.00%      51s 
Epoch 10:   89.74%      90.00%      51s 
Epoch 11:   89.74%      90.00%      51s 
Epoch 12:   89.74%      90.00%      51s 
Epoch 13:   89.74%      90.00%      51s 
Epoch 14:   89.74%      90.00%      51s
I am wonderring if any configuration is wrong.

**Comment by yueatsprograms** [comment]
Thanks for your interest in our work. Your configuration is correct, and I suspect that this is the third point in our Known Problems section of the README. We have never encountered this non-convergence through our many test runs but recognize this might be a problem for some people. You can try Xavier's solution. Please let me know it this helps.

**Comment by wlw208dzy** [comment]
Thanks for your reply again. I have tried the Xavier's solution in README, but it seems that the val result does not converge at all. As for the enviorment, Torch is the latest version and cudnn is v4. I followed the install intructions mentioned in https://github.com/facebook/fb.resnet.torch.

PS: After modifying the **main.lua** and **ResidualDrop.lua** as Xavier says, the val error began to converge immediately.

**Comment by yueatsprograms** [comment]
Would it converge for constant depth? Or for a shorter network with -N 3?

**Comment by wlw208dzy** [comment]
I have tried the suggestions you mentioned above, but the val result also would not converge. Maybe the problem I faced is a rare case because the training procedure on ImageNet dataset does not converge when using the code from https://github.com/facebook/fb.resnet.torch. My CUDA version is 7.0.18, so it is a CUDA issue possibly? I have tried to config the Torch environment for many times and different server machines. Some small networks are OK, but for deep residual networks on ImageNet dataset, the val errors are always abnormal. Need your help sincerely.

**Comment by yueatsprograms** [comment]
I'm sorry I don't have a better intuition for this kind of bug. Have you posted an issue on the Facebook code's github? Maybe they would have a better idea of what's going on?

I'm going to leave this thread open so other people can come across and help you.

**Comment by xgastaldi** [comment]
What worked for me is to modify the batchnorms in 3 places:

**In main.lua line 65:**
from: model:add(cudnn.SpatialBatchNormalization(16)) 
to: model:add(cudnn.SpatialBatchNormalization(16):init('weight', nninit.normal, 1.0, 0.002):init('bias', nninit.constant, 0))

**in ResidualDrop.lua lines 21 and 27**
from:  self.net:add(cudnn.SpatialBatchNormalization(nOutChannels))
to:   self.net:add(cudnn.SpatialBatchNormalization(nOutChannels):init('weight', nninit.normal, 1.0, 0.002):init('bias', nninit.constant, 0))

**Comment by wlw208dzy** [comment]
Oh, I found that I only modified the **ResidualDrop.lua**. So the BN layer before residual blocks initialized as default. After modifiying the **main.lua** just as Xavier says, the val error began to converge immediately. @yueatsprograms @xgastaldi Thanks for your help！

**Comment by wlw208dzy** [comment]
@xgastaldi Would you tell me why the initialization of BN layer has so much effect for the val result?

**Comment by yueatsprograms** [comment]
Would you mind to send

**Comment by xgastaldi** [comment]
Not exactly sure why but if I had to guess I would say that it maybe comes from too low initial batchnorm biases that put most of the data in the negative portions of the ReLUs

**Comment by yueatsprograms** [comment]
I have to guess this is a machine specific problem, somehow. The default initialization works very well when I test on our lab's servers, and for my friends who'v tried it. I'd really appreciate if someone finds out and let me know!

**Comment by ishann** [comment]
I experienced the same issue with a 110-layer Resnet on CIFAR-10.  Suggestion by @xgastaldi fixed the non-convergence. However, the Constant Depth model did not converge properly. Final test error is 7.37%. I've attached a plot of the errors obtained from running the code - [errorResnet110Cifar10.pdf](https://github.com/yueatsprograms/Stochastic_Depth/files/502290/errorResnet110Cifar10.pdf).
[Environment - Torch7, Ubuntu 14.04, CUDA 7.0.27, CuDNN R4]

Could any one suggest what I could try to ensure robust/consisten convergence as reported in the Stochastic Depth paper, as well as the first ResNet paper?
